oasislmf/pytools/common/data.py

```
@@ -141,3 +156,183 @@ def load_as_array(dir_path, name, _dtype, must_exist=True):
+    except Exception as e:
+        raise RuntimeError(f"An error occurred: {str(e)}")
+
+
+def read_periods(no_of_periods, run_dir, filename=PERIODS_FILE):
+    """Returns an array of period weights for each period between 1 and no_of_periods inclusive (with no gaps).
+    Args:
+        no_of_periods (int): Number of periods
+        run_dir (str | os.PathLike): Path to input files dir
+        filename (str | os.PathLike): periods binary file name
+    Returns:
+        period_weights (ndarray[period_weights_dtype]): Returns the period weights
+    """
+    periods_fp = Path(run_dir, filename)
+    period_weights_dtype = np.dtype([
+        ("period_no", np.int32),
+        ("weighting", np.float64),
+    ])
+
+    if not periods_fp.exists():
+        # If no periods binary file found, the revert to using period weights reciprocal to no_of_periods
+        logger.warning(f"Periods file not found at {periods_fp}, using reciprocal calculated period weights based on no_of_periods {no_of_periods}")
+        period_weights = np.array(
+            [(i + 1, 1 / no_of_periods) for i in range(no_of_periods)],
+            dtype=period_weights_dtype
+        )
+        return period_weights
+
+    try:
+        data = load_as_ndarray(run_dir, filename[:-4], period_weights_dtype)
+        # Less data than no_of_periods
+        if len(data) != no_of_periods:
+            raise RuntimeError(f"ERROR: no_of_periods does not match total period_no in period binary file {periods_fp}.")
+
+        # Sort by period_no
+        period_weights = np.sort(data, order="period_no")
+
+        # Identify any missing periods
+        expected_periods = np.arange(1, no_of_periods + 1)
+        actual_periods = period_weights['period_no']
+        missing_periods = np.setdiff1d(expected_periods, actual_periods)
+        if len(missing_periods) > 0:
+            raise RuntimeError(f"ERROR: Missing period_no in period binary file {periods_fp}.")
+    except Exception as e:
```

same, just let the error be raised